[CK TILE GEMM] Refactor block_scale_gemm examples

CongMa13 opened the pull request:
- Split cpp file to reduce building time
- Support multiple GemmConfig

## Proposed changes

The Gemm quant example faces two problems:
- Very long compile times.
- GemmConfig is hard-coded, which means you have to change the source and recompile to test different configurations.

In this PR, we'll split different instances into separate .cpp files to reduce compile time, and support multiple GemmConfig options selectable via command-line arguments.

- Move all template functions to run_gemm_quant_example.inc
- Add new cpp files to create gemm instances
- Add new command line options `data_type, quant_mode, preshuffleb, group_size` to select gemm instance
- Update main() 
  - Print help
  - Select GPU device

 

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [ ] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

CongMa13 committed:
[CK TILE GEMM] Refactor block_scale_gemm examples

- Split cpp file to reduce building time
- Support multiple GemmConfig

AviralGoelAMD commented on example/ck_tile/38_block_scale_gemm/gemm_utils.hpp:
```diff
@@ -11,12 +11,16 @@
+    size_t combined_hash = 0;
+    for(const auto& str : inputs)
+    {
+        combined_hash ^= hasher(str) + 0x9e3779b9 + (combined_hash << 6) + (combined_hash >> 2);
```
I had to google this (nice trick!), let us add a comment to make it more obvious.

```suggestion
        // Hash combine using golden ratio constant and bit shifts for good distribution and order-dependent mixing
        combined_hash ^= hasher(str) + 0x9e3779b9 + (combined_hash << 6) + (combined_hash >> 2);
```

CongMa13 replied:
done

AviralGoelAMD commented on example/ck_tile/38_block_scale_gemm/gemm_aquant_quantgrouped.cpp:
```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
+
+// This example demonstrates 2D block scale quantization (N×K) for BQuant
```
```suggestion
// This example demonstrates 2D block scale quantization (N×K) for BQuant
```
This comment is no longer relevant in this file now.

CongMa13 committed:
[CK TILE GEMM] Refactor block_scale_gemm examples

- Update Readme

CongMa13 commented on example/ck_tile/38_block_scale_gemm/run_gemm_quant_example.inc:
```diff
@@ -2,10 +2,233 @@
+            ck_tile::GemmPipelineAgBgCrCompV3<PipelineProblem>,
+            std::conditional_t<
+                QuantMode == ck_tile::QuantType::AQuantGrouped,
+                ck_tile::AQuantGemmPipelineAgBgCrCompV3<PipelineProblem>,
```
There are numerical errors with AQuantGemmPipelineAgBgCrMem

spolifroni-amd commented on example/ck_tile/38_block_scale_gemm/README.md:
```diff
@@ -40,23 +40,31 @@ This will result in an executable `build/bin/tile_example_gemm_quant_basic`
+      -stride_q    Tensor AQ stride (default:0)
+      -stride_b    Tensor B stride (default:0)
+      -stride_c    Tensor C stride (default:0)
+             -v    0. No validation, 1. Validation on CPU, 2. Validation on GPU (default:1)
```
For other options, the values and their descriptions are separated by colons. 

It should be this way here, too, not only for consistency, but also because this looks like an enumeration of options rather than a list of options and their descriptions.

```suggestion
             -v    0: No validation, 1: Validation on CPU, 2: Validation on GPU (default:1)
```

CongMa13 replied:
Updated

spolifroni-amd commented on example/ck_tile/38_block_scale_gemm/README.md:
```diff
@@ -40,23 +40,31 @@ This will result in an executable `build/bin/tile_example_gemm_quant_basic`
+      -stride_b    Tensor B stride (default:0)
+      -stride_c    Tensor C stride (default:0)
+             -v    0. No validation, 1. Validation on CPU, 2. Validation on GPU (default:1)
+          -prec    Data type. For AQuant: fp8/bf8/i4fp8/i4bf8, For Bquant: fp8/bf8/fp8i4/bf8i4 (default:fp8)
```
```suggestion
          -prec    Data type. For AQuant: fp8, bf8, 4fp8, or i4bf8;  for Bquant: fp8, bf8, fp8i4, or bf8i4 (default for both AQuant and Bquant: fp8)
```

The capitalization in AQuant and Bquant seem off. Is that how they're supposed to be capitalized?

CongMa13 replied:
Updated

spolifroni-amd commented on example/ck_tile/38_block_scale_gemm/README.md:
```diff
@@ -40,23 +40,31 @@ This will result in an executable `build/bin/tile_example_gemm_quant_basic`
+      -stride_c    Tensor C stride (default:0)
+             -v    0. No validation, 1. Validation on CPU, 2. Validation on GPU (default:1)
+          -prec    Data type. For AQuant: fp8/bf8/i4fp8/i4bf8, For Bquant: fp8/bf8/fp8i4/bf8i4 (default:fp8)
+        -warmup    Number of iterations before benchmark the kernel (default:50)
```
```suggestion
        -warmup    Number of iterations before benchmarking the kernel (default:50)
```

CongMa13 replied:
Updated

CongMa13 committed:
[CK TILE GEMM] Refactor block_scale_gemm examples

- Add support for rowcol and tensor GEMM operations

CongMa13 committed:
[CK TILE GEMM] Refactor block_scale_gemm examples

- Update README

CongMa13 committed:
[CK TILE GEMM] Refactor block_scale_gemm examples

- Set quant group size to (1, 1, 64) for targets excluding gfx950, where warp tile size (16, 16, 128) is incompatible.
